js/ai/src/model/middleware.ts

```diff
@@ -235,6 +240,171 @@ export function augmentWithContext(
   };
 }
 
+/**
+ * Options for the `retry` middleware.
+ */
+export interface RetryOptions {
+  /**
+   * The maximum number of times to retry a failed request.
+   * @default 3
+   */
+  maxRetries?: number;
+  /**
+   * An array of `StatusName` values that should trigger a retry.
+   * @default ['UNAVAILABLE', 'DEADLINE_EXCEEDED', 'RESOURCE_EXHAUSTED', 'ABORTED', 'INTERNAL']
+   */
+  statuses?: StatusName[];
+  /**
+   * The initial delay between retries, in milliseconds.
+   * @default 0
+   */
+  initialDelayMs?: number;
+  /**
+   * The factor by which the delay increases after each retry (exponential backoff).
+   * @default 1
+   */
+  backoffFactor?: number;
+  /**
+   * A callback to be executed on each retry attempt.
+   */
+  onError?: (error: Error, attempt: number) => void;
+}
+
+const DEFAULT_RETRY_STATUSES: StatusName[] = [
+  'UNAVAILABLE',
+  'DEADLINE_EXCEEDED',
+  'RESOURCE_EXHAUSTED',
+  'ABORTED',
+  'INTERNAL',
+];
+
+/**
+ * Creates a middleware that retries requests on specific error statuses.
+ *
+ * ```ts
+ * const { text } = await ai.generate({
+ *   model: googleAI.model('gemini-2.5-pro'),
+ *   prompt: 'You are a helpful AI assistant named Walt, say hello',
+ *   use: [
+ *     retry({
+ *       maxRetries: 2,
+ *       initialDelayMs: 1000,
+ *       backoffFactor: 2,
+ *     }),
+ *   ],
+ * });
+ * ```
+ */
+export function retry(options: RetryOptions = {}): ModelMiddleware {
+  const {
+    maxRetries = 3,
+    statuses = DEFAULT_RETRY_STATUSES,
+    initialDelayMs = 0,
+    backoffFactor = 1,
+    onError,
+  } = options;
```

Comment:
We should probably do some validation of values so we are not handing people a foot-gun (or DOS-attacker). Especially maxRetries. If someone decides that maxRetries = 100000 is good and backoffFactor defaults to 1 and initialDelayMs defaults to 0; Then that's a bunch of useless and unnecessary traffic...

Reply:
I think a well meaning but uninformed person might think the best option is to retry as quickly as possible until it succeeds... Maybe at least enforce initialDelay > 1 (so it's not 0 * backoff or an increasingly smaller fraction of backoff) and maxRetries <= 25? Or some "reasonable" limit?

Reply:
I'm not completely disagreeing with you, but this feels solidly as a P2 problem and solution requires a set of decisions that I currently don't have enough information about so hardcoding some limitations feels arbitrary, so I'm choosing to not make a decision on this right now. This is definitely not a blocker.